### PR TITLE
Short-Circuit NOOP Mapping Updates Earlier

### PR DESCRIPTION
**Pull request by original-brownbear**
No need to actually run noop mapping updates or create a new document mapper
if nothing has changed. Also, we can do the compressing of the updates off of the master update
thread and make the updates slightly less heavy that way.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 1]
```diff
@@ -271,16 +271,22 @@ public DocumentMapper merge(String type, CompressedXContent mappingSource, Merge
...
-        Mapping mapping = mergeMappings(this.mapper, incomingMapping, reason);
-        DocumentMapper newMapper = newDocumentMapper(mapping, reason);
-        if (reason == MergeReason.MAPPING_UPDATE_PREFLIGHT) {
+    private DocumentMapper mergeAndApplyMappings(String mappingType, CompressedXContent mappingSource, MergeReason reason) {
+        final DocumentMapper currentMapper = this.mapper;
+        if (currentMapper != null && currentMapper.mappingSource().equals(mappingSource)) {
+            return currentMapper;
+        }
+        synchronized (this) {
+            Mapping incomingMapping = parseMapping(mappingType, mappingSource);
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:280]: This can probably be outside the synchronized block as well?

**Review by romseygeek — APPROVED** [review]
LGTM, nice improvement @original-brownbear

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:280]: Technically I think so yes. But the parsing code does reference `this.mapper` in some spots so I decided to keep it this way for easy to understand correctness. Will try to look into a follow-up to this that cleans up the synchronization here a little :)

File: server/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingClusterStateUpdateRequest.java [hunk 2]
```diff
@@ -9,19 +9,22 @@
...
  */
 public class PutMappingClusterStateUpdateRequest extends IndicesClusterStateUpdateRequest<PutMappingClusterStateUpdateRequest> {
 
-    private final String source;
+    private final CompressedXContent source;
 
-    public PutMappingClusterStateUpdateRequest(String source) {
-        this.source = source;
+    public PutMappingClusterStateUpdateRequest(String source) throws IOException {
+        this.source = new CompressedXContent(source);
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingClusterStateUpdateRequest.java:24]: I thought about this, but it seemed like the wrong approach. Either, and this is true IMO, the compression cost is so tiny here that it doesn't matter really. Or if it does indeed matter, the better fix seems to be to just serialize the `PutMappingRequest` with the mapping in compressed form right away (thus making it smaller and easier to read from the wire as well).

**Comment by original-brownbear** [comment]
Thanks Alan & Henning!

**Pull request merged**